File: mcp-list/packages/EventAnalyzer/src/field_explainer.py
"""Field Explainer
解释字段含义和枚举值
"""

import json
from typing import Dict, Any, Optional


class FieldExplainer:
    """字段解释器"""

    def explain_field(
        self,
        field_name: str,
        field_definition: Dict[str, Any],
        show_enum: bool = True
    ) -> Dict[str, Any]:
        """
        解释字段含义

        Args:
            field_name: 字段名称
            field_definition: 字段定义
            show_enum: 是否显示枚举值

        Returns:
            字段解释信息
        """
        result = {
            "field": field_name,
            "type": field_definition.get("type", "UNKNOWN"),
            "tips": field_definition.get("tips", ""),
            "description": field_definition.get("desc", "")
        }

        # 解析枚举值
        if show_enum and field_definition.get("trans"):
            try:
                enum_values = json.loads(field_definition["trans"])
                result["enum_values"] = enum_values
                result["enum_count"] = len(enum_values)
            except json.JSONDecodeError:
                result["enum_values"] = {}

        return result

    def format_field_info(self, field_info: Dict[str, Any]) -> str:
        """
        格式化字段信息为易读的文本

        Args:
            field_info: 字段信息

        Returns:
            格式化后的文本
        """
        lines = []
        lines.append(f"字段名: {field_info['field']}")
        lines.append(f"类型: {field_info['type']}")
        lines.append(f"说明: {field_info['tips']}")

        if field_info.get("description"):
            lines.append(f"描述: {field_info['description']}")

        if field_info.get("enum_values"):
            lines.append("\\n枚举值:")
            for key, value in field_info["enum_values"].items():
                lines.append(f"  {key}: {value}")

        return "\\n".join(lines)

    def search_related_fields(
        self,
        field_name: str,
        all_fields: Dict[str, Dict]
    ) -> list[str]:
        """
        搜索相关字段（基于命名模式）

        Args:
            field_name: 字段名称
            all_fields: 所有字段定义

        Returns:
            相关字段名称列表
        """
        related = []

        # 提取字段名的主要部分
        parts = field_name.split("_")

        for other_field in all_fields.keys():
            if other_field == field_name:
                continue

            # 检查是否包含相同的词根
            other_parts = other_field.split("_")
            common_parts = set(parts) & set(other_parts)

            if len(common_parts) > 0:
                related.append(other_field)

        return related[:10]  # 最多返回 10 个相关字段

File: mcp-list/packages/EventAnalyzer-node/src/services/fieldExplainer.ts
/**
 * 字段解释器
 * 解释字段含义并搜索相关字段
 */

import { FieldDefinition, FieldDefinitions } from './apiClient.js';

export interface ExplanationResult {
  field_name: string;
  type: string;
  description: string;
  required: boolean;
  enum_values?: Record<string, string>;
  related_fields?: string[];
}

export class FieldExplainer {
  /**
   * 解释字段含义
   * @param fieldName 字段名称
   * @param fieldInfo 字段信息
   * @param showEnum 是否显示枚举值
   */
  explainField(fieldName: string, fieldInfo: FieldDefinition, showEnum: boolean = true): ExplanationResult {
    const result: ExplanationResult = {
      field_name: fieldName,
      type: fieldInfo.type,
      description: fieldInfo.desc || '无描述',
      required: fieldInfo.required || false,
    };

    // 解析枚举值
    if (showEnum && fieldInfo.trans) {
      result.enum_values = this.parseEnumValues(fieldInfo.trans);
    }

    return result;
  }

  /**
   * 搜索相关字段
   * 基于字段名称的相似度，返回相关字段列表
   * @param fieldName 目标字段名称
   * @param allFields 所有字段定义
   * @param maxResults 最大返回数量
   */
  searchRelatedFields(
    fieldName: string,
    allFields: FieldDefinitions,
    maxResults: number = 5
  ): string[] {
    const related: Array<{ field: string; score: number }> = [];

    for (const otherField in allFields) {
      if (otherField === fieldName) {
        continue;
      }

      const score = this.calculateSimilarity(fieldName, otherField);
      if (score > 0) {
        related.push({ field: otherField, score });
      }
    }

    // 按相似度排序
    related.sort((a, b) => b.score - a.score);

    // 返回前 N 个
    return related.slice(0, maxResults).map((item) => item.field);
  }

  /**
   * 计算两个字符串的相似度
   * 基于共同子串和长度
   */
  private calculateSimilarity(str1: string, str2: string): number {
    const s1 = str1.toLowerCase();
    const s2 = str2.toLowerCase();

    let score = 0;

    // 1. 完全包含关系
    if (s1.includes(s2) || s2.includes(s1)) {
      score += 5;
    }

    // 2. 共同前缀
    const commonPrefix = this.getCommonPrefix(s1, s2);
    if (commonPrefix.length > 2) {
      score += commonPrefix.length;
    }

    // 3. 共同后缀
    const commonSuffix = this.getCommonSuffix(s1, s2);
    if (commonSuffix.length > 2) {
      score += commonSuffix.length;
    }

    // 4. 共同子串
    const commonSubstrings = this.findCommonSubstrings(s1, s2);
    for (const substring of commonSubstrings) {
      if (substring.length > 2) {
        score += substring.length * 0.5;
      }
    }

    return score;
  }

  /**
   * 获取共同前缀
   */
  private getCommonPrefix(str1: string, str2: string): string {
    let i = 0;
    while (i < str1.length && i < str2.length && str1[i] === str2[i]) {
      i++;
    }
    return str1.substring(0, i);
  }

  /**
   * 获取共同后缀
   */
  private getCommonSuffix(str1: string, str2: string): string {
    let i = 0;
    while (
      i < str1.length &&
      i < str2.length &&
      str1[str1.length - 1 - i] === str2[str2.length - 1 - i]
    ) {
      i++;
    }
    return str1.substring(str1.length - i);
  }

  /**
   * 查找共同子串
   */
  private findCommonSubstrings(str1: string, str2: string): string[] {
    const substrings: Set<string> = new Set();

    for (let i = 0; i < str1.length; i++) {
      for (let j = i + 3; j <= str1.length; j++) {
        const substring = str1.substring(i, j);
        if (str2.includes(substring)) {
          substrings.add(substring);
        }
      }
    }

    return Array.from(substrings);
  }

  /**
   * 解析枚举值字符串
   */
  private parseEnumValues(transStr: string): Record<string, string> {
    const result: Record<string, string> = {};

    try {
      const items = transStr.split(',');
      for (const item of items) {
        const [key, value] = item.split(':');
        if (key && value) {
          result[key.trim()] = value.trim();
        }
      }
    } catch {
      // 解析失败
    }

    return result;
  }
}
